x/wasm/keeper/msg_server_integration_test.go

```diff
@@ -136,3 +141,643 @@ func TestUpdateParams(t *testing.T) {
+				m.InstantiatePermission = spec.permission
+			})
+
+			// store code
+			rsp, err := wasmApp.MsgServiceRouter().Handler(msg)(ctx, msg)
+			require.NoError(t, err)
+			var result types.MsgStoreCodeResponse
+			require.NoError(t, wasmApp.AppCodec().Unmarshal(rsp.Data, &result))
+
+			// when
+			msgInstantiate := &types.MsgInstantiateContract{
+				Sender: spec.addr,
+				Admin:  myAddress.String(),
+				CodeID: result.CodeID,
+				Label:  "test",
+				Msg:    []byte(`{}`),
+				Funds:  sdk.Coins{},
+			}
+			_, err = wasmApp.MsgServiceRouter().Handler(msgInstantiate)(ctx, msgInstantiate)
+
+			//then
+			if spec.expErr {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestInstantiateContract2(t *testing.T) {
+	wasmApp := app.Setup(t)
+	ctx := wasmApp.BaseApp.NewContext(false, tmproto.Header{Time: time.Now()})
+
+	var (
+		myAddress sdk.AccAddress = make([]byte, types.ContractAddrLen)
+		authority                = wasmApp.WasmKeeper.GetAuthority()
+	)
+
+	specs := map[string]struct {
+		addr       string
+		salt       string
+		permission *types.AccessConfig
+		expErr     bool
+	}{
+		"authority can instantiate a contract when permission is nobody": {
+			addr:       authority,
+			permission: &types.AllowNobody,
+			salt:       "salt1",
+			expErr:     false,
+		},
+		"other address cannot instantiate a contract when permission is nobody": {
+			addr:       myAddress.String(),
+			permission: &types.AllowNobody,
+			salt:       "salt2",
+			expErr:     true,
+		},
+		"authority can instantiate a contract when permission is everybody": {
+			addr:       authority,
+			permission: &types.AllowEverybody,
+			salt:       "salt3",
+			expErr:     false,
+		},
+		"other address can  instantiate a contract when permission is everybody": {
+			addr:       myAddress.String(),
+			permission: &types.AllowEverybody,
+			salt:       "salt4",
+			expErr:     false,
+		},
+	}
+	for name, spec := range specs {
+		t.Run(name, func(t *testing.T) {
+			// setup
+			_, _, sender := testdata.KeyTestPubAddr()
+			msg := types.MsgStoreCodeFixture(func(m *types.MsgStoreCode) {
+				m.WASMByteCode = wasmContract
+				m.Sender = sender.String()
+				m.InstantiatePermission = spec.permission
+			})
+
+			// store code
+			rsp, err := wasmApp.MsgServiceRouter().Handler(msg)(ctx, msg)
+			require.NoError(t, err)
+			var result types.MsgStoreCodeResponse
+			require.NoError(t, wasmApp.AppCodec().Unmarshal(rsp.Data, &result))
+
+			// when
+			msgInstantiate := &types.MsgInstantiateContract2{
+				Sender: spec.addr,
+				Admin:  myAddress.String(),
+				CodeID: result.CodeID,
+				Label:  "label",
+				Msg:    []byte(`{}`),
+				Funds:  sdk.Coins{},
+				Salt:   []byte(spec.salt),
+				FixMsg: true,
+			}
+			_, err = wasmApp.MsgServiceRouter().Handler(msgInstantiate)(ctx, msgInstantiate)
+
+			//then
+			if spec.expErr {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+			}
+		})
+	}
+}
```

Comment:
Very nice to see all these tests! This will give high confidence that things work as expected. Thanks for taking care!

Can you add a test for `UpdateInstantiateConfig` so that we have this covered, too?